File: code/003/prac03.R
setwd("/Users/mfjkou/sengokulab/study/awesomebook/")
install.packages("tidyverse")
library(tidyverse)

reserve_tb <- read.csv("data/reserve.csv", fileEncoding = "UTF-8",
                       header = TRUE, stringsAsFactors = FALSE)


#集計処理-カウント
#ホテル別の予約数と顧客数を知りたい
reserve_tb %>%
  group_by(hotel_id) %>%
  
  #summariseで集約処理を指定
  #n_functionでカウント処理, n_distinctでユニークに数え上げる
  summarise(rsv_cnt = n(),
            cus_cnt = n_distinct(customer_id))

#集計処理-合計
#宿泊人数別に売り上げ金額の合計を知りたい
reserve_tb %>%
  group_by(hotel_id, people_num) %>%
  summarise(earnings = sum(total_price)) %>%
  as.data.frame()

#集計処理-要約統計量
#ホテル別の金額の色々な情報を見たい
#quantile関数は下側分位点をとる
reserve_tb %>%
  group_by(hotel_id) %>%
  summarise(price_min = min(total_price),
            price_max = max(total_price), 
            price_avg = mean(total_price),
            price_median = median(total_price),
            price_20 = quantile(total_price, 0.2)) %>%
  as.data.frame()
  
#集計処理-分散と標準偏差
#ホテル別金額のばらつきを見る
#データ数1のときはNAになるのでこれを0とするのに注意

reserve_tb %>%
  group_by(hotel_id) %>%
  summarise(variance = coalesce(var(total_price), 0),
            st_deviation = coalesce(sd(total_price), 0)) %>%
  as.data.frame()
#別解
reserve_tb %>%
  group_by(hotel_id) %>%
  summarise(variance = var(total_price), st_deviation = sd(total_price)) %>%
  as.data.frame() %>%
  replace_na(list(variance=0, st_deviation = 0))


#集計処理-最頻値を発見する
#1000単位で四捨五入したあとに集計する-table関数
#カテゴリ別のカウントが最大であるものを見つける
#which.maxで最大値をもつベクトル情報を抽出し、namesで属性情報抽出
names(which.max(table(round(reserve_tb$total_price, -3))))


#集計処理-順序付
#比較のために時間型データに変換してからwindow関数を使う
reserve_tb$reserve_datetime <- as.POSIXct(reserve_tb$reserve_datetime, format = '%Y-%m-%d %H:%M:%S')

reserve_tb %>%
  group_by(customer_id) %>%
  mutate(log_no = row_number(reserve_datetime)) %>%
  as.data.frame()

#ホテルごとの予約数を出して, その予約数で順位付する

reserve_tb %>%
  group_by(hotel_id) %>%
  summarise(rsv_cnt = n()) %>%
  #予約回数をもとに順位を算出する
  #同じランクなら最小順位を全てつけてあげる->min_rank + descで降順
  #transmuteで指定した列のみ残す
  transmute(hotel_id, rsv_cnt_rank = min_rank(desc(rsv_cnt))) %>%
  as.data.frame()

File: code/003/prac03.py
#%%
import numpy as np 
import pandas as pd 
df = pd.read_csv('data/reserve.csv', encoding = 'UTF-8')

#%%
# 集計処理を行う
#ホテルの別の予約数と顧客数を知りたい
#size関数とnunique関数で集計, agg関数で同時処理 

#%%
result = df.groupby('hotel_id') \
    .agg({'reserve_id':'size', 'customer_id':'nunique'})
result.reset_index(inplace = True)
result.columns = ['hotel_id', 'rsv_cnt', 'cus_cnt']

#%%
result.head(3)

#%%
result = df.groupby(['hotel_id', 'people_num'])['total_price'] \
    .sum().reset_index()
result.rename(columns = {'total_price':'eanring_sum'}, inplace=True)
#列名の書き換えが少ない時はrenameの方が便利

#%%
result.head(5)

#%%
#ホテル別宿泊金額の要約統計量
#aggの中にlambda式を入れて分位点も出せる
result = df.groupby('hotel_id')['total_price'] \
    .agg({'total_price':['min', 'max',
    'mean', 'median', lambda x: np.percentile(x, q=20)]}) \
    .reset_index() #np.quantileだとno moduleになる. なぜ？
result.columns = ['hotel_id','price_min', 'price_max', 'price_avg', 'price_median', 'price_20']

#%%
result.head()

#%%
#分散の計算
result = df.groupby('hotel_id')['total_price'] \
.agg({'total_price': ['var', 'std']}).reset_index()
result.columns = ['hotel_id', 'variance', 'st_deviation']
result.fillna(value = {'variance':0, 'st_deviation':0}, inplace = True)

#%%
result.head()

#%%
#1000単位の最頻値を算出する-mode関数でOK
df['total_price'].round(-3).mode()

#%%
#順序付処理 #顧客ごとの予約で時間ごとに順位付する
df['reserve_datetime'] = pd.to_datetime(df['reserve_datetime'])
df['log_no'] = df.groupby('customer_id')['reserve_datetime'] \
        .rank(ascending=True, method = 'first')

#%%
df.head(20)

#%%
#順序付け処理その2 #ホテルごとの予約数を算出して順位付
rsv_cnt_tb = df.groupby('hotel_id').size().reset_index()
rsv_cnt_tb.columns = ['hotel_id', 'rsv_cnt']


#%%
rsv_cnt_tb['rsv_cnt_rank'] = rsv_cnt_tb['rsv_cnt'].rank(ascending = True, method = 'min')
rsv_cnt_tb.drop('rsv_cnt', axis=1, inplace=True)

#%%
rsv_cnt_tb.head()

#%%
